```
error: builder for '/nix/store/m24vv0aga6fqcvpam66wkv9153rsyb91-postgis-aarch64-unknown-linux-gnu-3.5.3.drv' failed with exit code 1;
       last 25 log lines:
       > checking for CUNIT... no
       > configure: WARNING: could not locate CUnit required for unit tests
       > checking for iconv.h... yes
       > checking for libiconv_open in -liconv... no
       > checking for iconv_open in -lc... yes
       > checking for iconvctl... no
       > checking for libiconvctl... no
       > Using user-specified pg_config file: /nix/store/r1ln26jp7whqszz2jvfzrxj1r8gq4g1y-pg_config/bin/pg_config
       > checking PostgreSQL version... PostgreSQL 17.5
       > checking for libpq-fe.h... yes
       > checking for PQserverVersion in -lpq... yes
       > ------------------------------------------------------------------------
       >   WARNING: You have set the --prefix to '/nix/store/7s4sxcxjzhb9hgkwdl8g69kcyhdjjkxj-postgis-aarch64-unknown-linux-gnu-3.5.3'. But we mostly
       >   ignore the --prefix. For your info, using the values determined from
       >   /nix/store/r1ln26jp7whqszz2jvfzrxj1r8gq4g1y-pg_config/bin/pg_config we will be installing:
       >     * postgis shared library in /nix/store/gbalc300kv6dn7dpvanfh6yr31cqaikj-postgresql-aarch64-unknown-linux-gnu-17.5-lib/lib
       >     * postgis SQL files in /nix/store/qw05cl957d0axvldjwl9fqa340i7fxxg-postgresql-aarch64-unknown-linux-gnu-17.5/share/postgresql/contrib/postgis-3.5
       > ------------------------------------------------------------------------
       > checking for xml2-config... /nix/store/j2gi38k4h3hr2mrqaqisv774p2kypyqm-libxml2-2.14.4-unstable-2025-06-20-dev/bin/xml2-config
       > checking for libxml/tree.h... yes
       > checking for libxml/parser.h... yes
       > checking for libxml/xpath.h... yes
       > checking for libxml/xpathInternals.h... yes
       > checking for xmlInitParser in -lxml2... no
       > configure: error: could not find libxml2
```

--- a/pkgs/servers/sql/postgresql/ext/postgis.nix
+++ b/pkgs/servers/sql/postgresql/ext/postgis.nix
@@ -91,12 +91,19 @@ postgresqlBuildExtension (finalAttrs: {
     ./autogen.sh
   '';
 
-  configureFlags = [
-    "--with-pgconfig=${postgresql.pg_config}/bin/pg_config"
-    "--with-gdalconfig=${gdal}/bin/gdal-config"
-    "--with-jsondir=${json_c.dev}"
-    "--disable-extension-upgrades-install"
-  ] ++ lib.optional withSfcgal "--with-sfcgal=${sfcgal}/bin/sfcgal-config";
+  configureFlags =
+    let
+      isCross = stdenv.hostPlatform.config != stdenv.buildPlatform.config;
+    in
+    [
+      (lib.enableFeature false "extension-upgrades-install")
+      (lib.withFeatureAs true "pgconfig" "${postgresql.pg_config}/bin/pg_config")
+      (lib.withFeatureAs true "gdalconfig" "${gdal}/bin/gdal-config")
+      (lib.withFeatureAs true "jsondir" (lib.getDev json_c))
+      (lib.withFeatureAs true "xml2config" (lib.getExe' (lib.getDev libxml2) "xml2-config"))
+      (lib.withFeatureAs withSfcgal "sfcgal" "${sfcgal}/bin/sfcgal-config")
+      (lib.withFeature (!isCross) "json") # configure: error: cannot check for file existence when cross compiling
+    ];
 
   makeFlags = [
     "PERL=${perl}/bin/perl"

